App/EnclaveManager.cpp

```diff
@@ -1,35 +1,93 @@
-#include <stdio.h>
-#include "EnclaveManager.h"
-#include "SgxException.h"
-#include "Enclave_u.h"
-#include "ErrorCode.h"
-
-EnclaveManager::EnclaveManager() {
-	sgx_enclave_id_t eid;
-	sgx_status_t ret = SGX_SUCCESS;
-	sgx_launch_token_t token = { 0 };
-	int updated = 0;
-
-	ret = sgx_create_enclave(ENCLAVE_FILENAME, SGX_DEBUG_FLAG,
-		&token, &updated,
-		&eid, NULL);
-
-	if (ret != SGX_SUCCESS) throw SgxException(ret);
-	this->eid = eid;
-}
-
-EnclaveManager::~EnclaveManager() {
-	sgx_status_t ret = SGX_SUCCESS;
-	ret = sgx_destroy_enclave(eid);
-
-	if (ret != SGX_SUCCESS) throw SgxException(ret);
-}
-
-unsigned long EnclaveManager::generateRandom() {
-	unsigned int ret = FC_OK;
-	unsigned long result;
-	e_generate_random_long(eid, &ret, &result);
-	if (ret != FC_OK) throw SgxException(SGX_ERROR_UNEXPECTED);
-
-	return result;
-}
+#include "EnclaveManager.h"
+#include "Attestator.h"
+#include "SgxException.h"
+#include "Enclave_u.h"
+#include "ErrorCode.h"
+
+#include <sgx_report.h>
+#include <sgx_uae_service.h>
+#include <sgx_ukey_exchange.h>
+#include <sgx_utils.h>
+#include <time.h>
+#include <string>
+#include <cstdint>
+#include <vector>
+#include <iostream>
+#include <stdio.h>
+#include <sys/stat.h>
+
+using std::vector;
+using std::string;
+using std::to_string;
+using std::invalid_argument;
+
+#define ENCLAVE_NAME "libs/Enclave.signed.so"
+
+EnclaveManager::EnclaveManager() {
+	sgx_enclave_id_t eid;
+	sgx_status_t ret = SGX_SUCCESS;
+	sgx_launch_token_t token = { 0 };
+	int updated = 0;
+
+	ret = sgx_create_enclave(ENCLAVE_FILENAME, SGX_DEBUG_FLAG,
+		&token, &updated,
+		&eid, NULL);
+
+	if (ret != SGX_SUCCESS) throw SgxException(ret);
+	this->eid = eid;
+}
+
+EnclaveManager::~EnclaveManager() {
+	sgx_status_t ret = SGX_SUCCESS;
+	ret = sgx_destroy_enclave(eid);
+
+	if (ret != SGX_SUCCESS) throw SgxException(ret);
+}
+
+unsigned long EnclaveManager::generateRandom() {
+	unsigned int ret = FC_OK;
+	unsigned long result;
+	enclave_generate_random_long(eid, &ret, &result);
+	if (ret != FC_OK) throw SgxException(SGX_ERROR_UNEXPECTED);
+
+	return result;
+}
+
+unsigned int EnclaveManager::initRemoteAttestation() {
+    config_t config;
+	sgx_launch_token_t token= { 0 };
+	sgx_status_t status;
+	sgx_enclave_id_t eid= 0;
+	int updated= 0;
+	int sgx_support;
+	uint32_t i;
+	EVP_PKEY *service_public_key= NULL;
+	char have_spid= 0;
+	char flag_stdio= 0;
+
+	const time_t timeT = time(NULL);
+	struct tm lt, *ltp;
+
+	ltp = localtime(&timeT);
+	if ( ltp == NULL ) {
+		perror("localtime");
```

Comment:
In some places you're throwing exceptions and in the other return statuses with some logging. I think it would be better to handle all errors the same way.

Reply:
This is because I was adapting to Intel code that has return statuses. I will keep Attestation/* files as they are, but apply your suggestion in App/* files.